webui/src/components/CheckpointDetails.tsx

```diff
@@ -30,68 +57,228 @@ function formatDuration(micros: number): string {
   }
 }
 
-const spanDuration = (subtask: SubtaskCheckpointGroup, spanType: CheckpointSpanType) => {
-  const span = subtask.eventSpans.find(s => s.spanType == spanType);
-  return span ? formatDuration(span.finishTime - span.finishTime) : 'n/a';
-};
+const CheckpointTimeline: React.FC<CheckpointTimelineProps> = ({ operators, checkpoint }) => {
+  const svgRef = useRef<SVGSVGElement>(null);
+  const tooltipRef = useRef<HTMLDivElement>(null);
 
-const spans = (subtasks: SubtaskCheckpointGroup[], spanType: CheckpointSpanType) => {
-  return (
-    <Stack>
-      {subtasks.map(s => (
-        <Text>{spanDuration(s, spanType)}</Text>
-      ))}
-    </Stack>
-  );
-};
+  useEffect(() => {
```

Comment:
This useEffect seems heavy. Can the graph structure and the data be separated into two things?